deepchem/data/datasets.py

```diff
@@ -1574,6 +1589,55 @@ def iterate(dataset: DiskDataset, batch_size: Optional[int], epochs: int):
 
     return iterate(self, batch_size, epochs)
 
+  def _get_shard_for_index(self, index: int) -> Tuple[int, int]:
+    """Get the shard and shard index for a particular datapoint."""
+    if self.legacy_metadata:
+      raise ValueError(
+          "Indexing not supported for legacy metadata. Please convert metadata by calling self.reshard()"
+      )
+    if index < 0 or index >= len(self):
+      raise IndexError("Index %d is out of bounds" % index)
```

Comment:
This check is very expensive.  At best, `len()` involves a loop over all shards.  And if the dataset has legacy metadata, it involves actually loading all the ids from disk for the entire dataset.  You don't want to repeat that every time the user accesses one sample.